internal/gojs/wasm/wasm.go

```diff
@@ -0,0 +1,526 @@
+// Package wasm are shared code related to runtime.GOARCH "wasm".
+package wasm
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/internal/gojs/js"
+	"github.com/tetratelabs/wazero/internal/wasm"
+)
+
+var le = binary.LittleEndian
+
+// refreshSP refreshes the stack pointer, which is needed prior to storeValue
+// when in an operation that can trigger a Go event handler.
+//
+// See https://github.com/golang/go/blob/go1.19/misc/wasm/wasm_exec.js#L210-L213
+func refreshSP(mod api.Module) uint32 {
+	// Cheat by reading global[0] directly instead of through a function proxy.
+	// https://github.com/golang/go/blob/go1.19/src/runtime/rt0_js_wasm.s#L87-L90
+	return uint32(mod.(*wasm.CallContext).GlobalVal(0))
+}
+
+func mustReadN(mem api.Memory, sp, n uint32) []byte {
+	return mustRead(mem, "sp", sp+8, n<<3)
+}
+
+// mustRead is like api.Memory except that it panics if the offset and
+// byteCount are out of range.
+func mustRead(mem api.Memory, fieldName string, offset, byteCount uint32) []byte {
+	buf, ok := mem.Read(offset, byteCount)
+	if !ok {
+		panic(fmt.Errorf("out of memory reading %s", fieldName))
+	}
+	return buf
+}
+
+func newFunc(name string, goFunc api.GoModuleFunc) *wasm.HostFunc {
+	return &wasm.HostFunc{
+		ExportNames: []string{name},
+		Name:        name,
+		ParamTypes:  []api.ValueType{api.ValueTypeI32},
+		ParamNames:  []string{"sp"},
+		Code:        &wasm.Code{IsHostFunction: true, GoFunc: goFunc},
+	}
+}
+
+// StubFunction stubs functions not used in Go's main source tree.
+// This traps (unreachable opcode) to ensure the function is never called.
+func StubFunction(name string) *wasm.HostFunc {
+	return &wasm.HostFunc{
+		ExportNames: []string{name},
+		Name:        name,
+		ParamTypes:  []wasm.ValueType{wasm.ValueTypeI32},
+		ParamNames:  []string{"sp"},
+		Code:        &wasm.Code{IsHostFunction: true, Body: []byte{wasm.OpcodeUnreachable, wasm.OpcodeEnd}},
+	}
+}
+
+func NoopFunction(name string) *wasm.HostFunc {
+	return &wasm.HostFunc{
+		ExportNames: []string{name},
+		Name:        name,
+		ParamTypes:  []wasm.ValueType{wasm.ValueTypeI32},
+		ParamNames:  []string{"sp"},
+		Code:        &wasm.Code{IsHostFunction: true, Body: []byte{wasm.OpcodeEnd}},
+	}
+}
+
+func NewFuncR_V(name string, goFunc FuncR_V, pName string) *wasm.HostFunc {
```

Comment:
from here down handle all the go signatures used by `GOARCH=wasm GOOS=js` derived from the SP parameter. There won't be more (I think) as inside the javascript "operating system" "valueCall" uses dynamic dispatch, which doesn't affect the stack except arg length.